File: Algorithms/Q13 - Roman to Integer/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function (s) {
  let num = 0;
  let dict = {
    I: 1,
    V: 5,
    X: 10,
    L: 50,
    C: 100,
    D: 500,
    M: 1000,
  };
  let charCanBeMinus = {
    C: ["M", "D"],
    X: ["C", "L"],
    I: ["V", "X"],
  };
  for (let index = 0; index < s.length; index++) {
    if (
      s[index] in charCanBeMinus &&
      index < s.length - 1 &&
      charCanBeMinus[s[index]].includes(s[index + 1])
    ) {
      num += dict[s[index + 1]] - dict[s[index]];
      index += 1;
    } else {
      num += dict[s[index]];
    }
  }
  return num;
};

File: Algorithms/Q13 - Roman to Integer/solution.java
class Solution {
    public int romanToInt(String s) {
        int num = 0;
        HashMap<Character, Integer> dict = new HashMap<Character, Integer>() {{
            put('I', 1);
            put('V', 5);
            put('X', 10);
            put('L', 50);
            put('C', 100);
            put('D', 500);
            put('M', 1000);
        }};
        for (int index = 0; index < s.length(); index++) {
            if (index < s.length() - 1 && dict.get(s.charAt(index)) < dict.get(s.charAt(index + 1)) ) {
                num += dict.get(s.charAt(index + 1)) - dict.get(s.charAt(index));
                index += 1;
            } else {
                num += dict.get(s.charAt(index));
            }
        }
        return num;
    }
}